### PR TITLE
Modules View

### LINKED ISSUE #101: Modules Frontend Component
image
^ here is the image of the view to create 

In order to get a sense of the API routes, take a look at #30 . It has a good overview of how to post and add modules and what they contain (items) into the database. Adding a module and item shouldnt be too bad - both just require interacting with some POST routes. 

However, the best way to go about this is probably creating a component called `Module` that renders the dropdown and all the items within it. Within the modules component, you can make a call to the backend to get all the links that belong within that module. This component should also handle the adding, deleting, and editing of the individual items. Creating an item should just have a pop up that looks like the create module pop up, but with name and link instead. The API does not have support to editing modules, so if you could add that in, that would be great. Should just be a PATCH to `/api/module/[module_id]/item/[item_id]` with the following at the body:
```
type EditItem = {
title?: string, 
link?: string 
}
```

Finally, make sure this modules parameter takes in some prop called `enableEditing`. Only allow editing functions to appear when this is set to `true`

In the overall page, we can make a call to `GET /api/class/[id]/modules
` to get all the modules that belong to that class. Then, simply render the `Modules` components in the order specified. 

We can add the ability to move a module up and down. As soon as someone asks a module to move up, we can fire off a call to the database with the new order. We can do this by having an array of moudle id’s in the order they are currently being displayed, and having a function that edits the order when one of the move up/down buttons are called. Make sure to keep track of the fact that the topmost module can’t move up and the bottom most module cant move down (maybe have a min/max that you check).

### PR DESCRIPTION
**Pull request by nishantbalaji**
### Administrative Info

Make sure your branch name conforms to: `<feature/staging/hotfix/...>/[username]/[Github Issue]-[3-4 word description separated by dashes]`.

### What issue(s) does this branch close?
Closes #101 

### Changes

What changes did you make?

- Added a Module view with dropdowns showing the modules

### Testing

How did you confirm your changes worked?

- Test with dummy data and make sure they dsplay

### Confirmation of Change

Upload a screenshot, if possible. Otherwise, please provide instructions on how to see the change.

| Module View |
| ----------------------- |
| image  |

### CODE REVIEW

**Review by Anshul-Birla — COMMENTED** [review]
The styling here has not been applied: 

image

File: components/SingleClassView/Module/Module.tsx [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            type="text"
+            placeholder="Lesson Title"
+          />
+          <input
+            className={`${styles.label} ${styles.classInput}`}
+            value={link}
+            onChange={(e) => setLink(e.target.value)}
+            type="text"
+            placeholder="Lesson Link"
+          />
```
Comment (Anshul-Birla) [inline components/SingleClassView/Module/Module.tsx:182]: you can make this a popup btw - it doesnt need to be an edit here

File: components/SingleClassView/Module/Module.tsx [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    </Accordion>
+  );
+};
+
+export const ClassModule: React.FC<ModuleProps> = ({ id }) => {
+  const api = useContext(APIContext);
+  const [popup, setPopup] = useState(false);
+  const [name, setName] = useState("");
+
+  const { data: modules, error } = useSWR(`api/class/${id}/modules`, () => api.getClassModules(id));
```
Comment (Anshul-Birla) [inline components/SingleClassView/Module/Module.tsx:228]: lets not use swr for this, its probably better to use `useEffect`

Comment (Anshul-Birla) [reply components/SingleClassView/Module/Module.tsx:228]: Take a look at this: https://github.com/TritonSE/WA-Family-Literacy-Application/blob/dc893e77fb15e8ded0a794835f338991b2e96fe3/web/src/pages/UploadBooksPage.tsx#L34

File: context/LeagueAPI.ts [hunk 3]
```diff
@@ -146,6 +146,32 @@ class LeagueAPI {
...
     return res.data;
   }
+
+  async getClassModules(classId: string): Promise<Module[]> {
+    const res = await this.client.get(`api/class/${classId}/modules`);
+    return res.data;
+  }
+
+  // eslint-disable-next-line
+  async getModuleItems(moduleId: string): Promise<any> {
```
Comment (Anshul-Birla) [inline context/LeagueAPI.ts:156]: reminder to fix the any

File: pages/api/module/[id]/item/[item_id].ts [hunk 4]
```diff
@@ -68,6 +70,24 @@ export const deleteItemHandler: NextApiHandler = async (
         return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json("Internal Server CustomError");
```
Comment (Anshul-Birla) [inline pages/api/module/[id]/item/[item_id].ts:70]: lets remover this custom error and just say internal server error

**Review by Anshul-Birla — CHANGES_REQUESTED** [review]
Some logical changes do need to be made, but i think the overall structure is correct

File: components/SingleClassView/Module/Module.tsx [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+import Menu from "@mui/material/Menu";
+import MenuItem from "@mui/material/MenuItem";
+import Fade from "@mui/material/Fade";
+
+type FadeMenuProps = {
+  module: Module;
+  numModules: number;
+};
+
+const FadeMenu: React.FC<FadeMenuProps> = ({ module, numModules }) => {
```
[the remark this answers is not in the document]
Comment (nishantbalaji) [reply components/SingleClassView/Module/Module.tsx:22]: ye

File: components/SingleClassView/Module/Module.tsx [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+interface APIModuleItem {
+  itemId: string;
+  link: URL;
+  moduleId: string;
+  title: string;
+}
+
+// eslint-disable-next-line
+const AccordionLesson = ({ lesson }: { lesson: APIModuleItem }) => {
```
[the remark this answers is not in the document]
Comment (nishantbalaji) [reply components/SingleClassView/Module/Module.tsx:135]: wait theyre two different things -- lesson is the thing inside the module

[the remark this answers is not in the document]
Comment (Anshul-Birla) [reply components/SingleClassView/Module/Module.tsx:135]: lets make it AccordionItem or something - lesson is too constrained of a name

**Review by Anshul-Birla — CHANGES_REQUESTED** [review]
Some functionality that is currently not working: 
- The image to edit does not show up:  
image
- When creating a module, the page does not auto-refresh to display the new module. 
- There is some styling issue on the side of the module components
image
- We shouldn't have borders around these boxes
image
- Hitting the 3 dots causes the entire component to expand. This should not be expected functionality 
- Creating an item does not automatically make it appear in the module dropdown 
- When in the create item page, the "cancel" button causes the module to just move up and down instead of cancelling

**Comment by nishantbalaji** [comment]
> Some functionality that is currently not working:
> 
> * The image to edit does not show up:
>   image
> * When creating a module, the page does not auto-refresh to display the new module.
> * There is some styling issue on the side of the module components
>   image
> * We shouldn't have borders around these boxes
>   image
> * Hitting the 3 dots causes the entire component to expand. This should not be expected functionality
> * Creating an item does not automatically make it appear in the module dropdown
> * When in the create item page, the "cancel" button causes the module to just move up and down instead of cancelling

@Anshul-Birla 

this is mostly fixed, however i need to fix the Vertical Menu toggling the component. I am not sure how to do this with MUI Accordion but I sent you an alternative via slack if that also works?

**Pull request closed without merging**